apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,378 @@
+// Each StagedUpdateRun object corresponds to a single "release" of a certain version of the resources.
+// The release is abandoned if the StagedUpdateRun object is deleted or the scheduling decision changes.
+type StagedUpdateRun struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The desired state of StagedUpdateRun. The spec is immutable.
+	// +immutable
+	// +required
+	Spec StagedUpdateRunSpec `json:"spec"`
+
+	// The observed status of StagedUpdateRun.
+	// +optional
+	Status StagedUpdateRunStatus `json:"status,omitempty"`
+}
+
+// StagedUpdateRunSpec defines the desired the update strategy and the snapshot indices
+// on which the update is based.
+type StagedUpdateRunSpec struct {
+	// A reference to the placement that this update run is applied to.
+	// There can be multiple active update runs for each placement but
+	// it's up to the devOps to make sure they don't conflict with each other.
+	// +required
+	PlacementRef PlacementReference `json:"placementRef"`
+
+	// The resource snapshot index that this update run is based on.
+	// +required
+	ResourceSnapShotIndex string `json:"resourceSnapShotIndex"`
+
+	// The override snapshot index that this update run is based on.
+	// +required
+	OverrideSnapShotIndex string `json:"overrideSnapShotIndex"`
+
+	// The reference to the update strategy that specifies the stages and the sequence
+	// in which the application will be updated on the member clusters.
+	// +required
+	StagedUpdateStrategyRef StagedUpdateStrategyReference `json:"stagedRolloutStrategyRef"`
+}
+
+// PlacementReference is a reference to a placement object.
+type PlacementReference struct {
+	// Name is the name of the referenced StagedUpdateStrategy.
+	Name string `json:"name"`
+	// Namespace is the namespace where the referenced StagedUpdateStrategy is located.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// StagedUpdateStrategyReference is a reference to a StagedUpdateStrategy.
+type StagedUpdateStrategyReference struct {
+	// Name is the name of the referenced StagedUpdateStrategy.
+	Name string `json:"name"`
+	// Namespace is the namespace where the referenced StagedUpdateStrategy is located.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// +genclient
+// +genclient:namespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Namespaced",categories={fleet,fleet-placement}
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// StagedUpdateStrategy defines a reusable strategy that specifies the stages and the sequence
+// in which the application will be updated on the member clusters.
+type StagedUpdateStrategy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Stage configurations for each update stage.
+	// +required
+	Stages []StageConfig `json:"stages"`
```

Comment:
Shall we have a type for the strategy as it may have more fields beyond stages? You can use the same type in UPdateRunStatus, rather than copying N fields to status. 

type XXX {
    maxUnavailable *int64
    Stages []StageConfig
}

Reply:
Not resolved?

Reply:
I am not sure what other type of strategy we will add, I think we can just add new fields in the StagedUpdateStrategy